agent/secrets/secrets.go

```diff
@@ -50,46 +51,37 @@ func (s *secrets) remove(secrets []string) {
 }
 
 // reset removes all the secrets
-func (s *secrets) reset() {
+func (s *secrets) Reset() {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.m = make(map[string]*api.Secret)
 }
 
-func (s *secrets) filter(secretIDs []string) map[string]*api.Secret {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	filteredSecrets := make(map[string]*api.Secret)
-
-	for _, secretID := range secretIDs {
-		if s, ok := s.m[secretID]; ok {
-			filteredSecrets[secretID] = s.Copy()
-		}
-	}
-
-	return filteredSecrets
+// taskRestrictedSecretsProvider restricts the ids to the task.
+type taskRestrictedSecretsProvider struct {
+	secrets   exec.SecretGetter
+	secretIDs map[string]struct{} // allow list of secret ids
 }
 
-// getStore returns ta Store with only the secrets corresponding to the IDs
-// that are passed in.
-func (s *secrets) getStore(secretIDs []string) exec.SecretProvider {
-	return &secrets{
-		m: s.filter(secretIDs),
+func (sp *taskRestrictedSecretsProvider) Get(secretID string) *api.Secret {
+	if _, ok := sp.secretIDs[secretID]; !ok {
+		return nil
 	}
+
+	return sp.secrets.Get(secretID)
 }
 
-// getStoreForTask returns only the secrets needed by a specific Task
-func (s *secrets) getStoreForTask(task *api.Task) exec.SecretProvider {
-	var secretIDs []string
+// Restrict provides a getter that only allows access to the secrets
+// referenced by the task.
+func Restrict(secrets exec.SecretGetter, t *api.Task) exec.SecretGetter {
```

Comment:
Why would we restrict by adding an ACL vs simply not passing the secrets in the first place?